meriadec/ledger-live-desktop
# Secure update

<p align="center">
  <img src="https://user-images.githubusercontent.com/315259/49865344-00696980-fe05-11e8-9a18-6cc0ede766ee.jpg" width="100" />
</p>

### Implementing secure updates

The purpose of this PR is to improve security of the update process. Two axis are involved:

- the update patch is checked internally just after being downloaded:
  - sha512sum check against sum file, stored outside of github
  - signature verification of sum file, with Ledger pubkey (embedded inside the app)
- the update doesn't install automatically anymore: a user action is required (click on "Install")

It also brings some refactoring on front side:

- `autoUpdate` check & events, and `quitAndInstall` action has been migrated to "command" pattern. Because they require `electron` dependency, there is now a way to launch a command in the main process instead of the internal process (afb8000eaee306d093b89b3a0331193472c2efeb). It should not affect existing behaviour in any way.
- `updater` reducer has been dropped in favor of `React.Context` pattern for multiple reasons — could develop later —, and multiple consumer components. As a side-effect, React has been bumped because of bug in context refreshing.

As a side effect, it colocates sha512sum of the binaries + their signatures next to the release binaries, which resolves #942.

**How the app handle a Ledger pubkey update?**

The particular scenario here, is if by any chance the public key has to change (it will be a qorum based key, so if any of the qorum people leave, a new key needs to be generated).

If we stay in the naive approach, the next updates will be refused. So the need is to be able to sign app with another key, sign the new key with the previous one, and make the updater verify everything.

```
                      +------------------------+   +--------------------+
                      | KEY SERVER             |   | GITHUB             |
                      |                        |   |                    |
                      | pubkeys/root.asc       |   | ll-1.8.0.AppImage  |
                      | pubkeys/key-2.asc      |   | ll-1.8.0.exe       |
                      | pubkeys/key-2.asc.sig  |   | ll-1.8.0.dmg       |
                      | pubkeys/key-3.asc      |   | ...                |
                      | pubkeys/key-3.asc.sig  |   +---+----------------+
                      | ll-1.8.0.sha512sum     |       |
                      | ll-1.8.0.sha512sum.sig |       |
                      | ll-1.9.0.sha512sum     |       |
                      | ll-1.9.0.sha512sum.sig |       |
                      | ...                    |       |
                      +------------+-----------+       |
                                   |                   |
                                   |     +-------------+
                                   |     |
                                   V     V
                                +--------------+
                                | LIVE DESKTOP |
                                |              |
                                | key-2.asc    |
                                +--------------+

1- check is made with embedded pubkey
2- if check fail, app try to fetch the next key, which is named
   with the fingerprint of the previous one, and signed with the previous one
3- if no key found, error
4- else, jump to step 1
```

**What's next?**

- [ ] I've setup a key-server example here: https://ledger-live.sigsev.io, but we need to **setup it inside our infra**. And **craft some scripts** to feed it with current release shasums, etc. And set the feed URL inside the Live ([here](https://github.com/meriadec/ledger-live-desktop/blob/2751d88cca433d532afb4c15dfda238115cebd37/src/config/constants.js#L101)).
- [ ] The actual [pubkey](https://github.com/meriadec/ledger-live-desktop/blob/secure-update/src/main/updater/ledger-pubkey.js) stored in source code is mine. We need to **put the qorum pubkey**.
- [ ] The check is only informative for us for the moment ("white bullet"). This means if the check fail, update is still marked as available (no false positive, etc.). This behaviour can be deactivated by changing the boolean [here](https://github.com/meriadec/ledger-live-desktop/blob/2751d88cca433d532afb4c15dfda238115cebd37/src/config/constants.js#L100).
- [ ] The update notification banner has had some changes, and require new wording. cf screenshot below. Could also need cosmetic polish. See screenshots at the end.

<p> </p>
<p> &nbsp;</p>
<p align="center">
  <img src="https://user-images.githubusercontent.com/315259/49885015-8225bb00-fe36-11e8-9ae9-e63b44a48b0f.jpg" width="100" />
</p>

### Test plan

All my tests were done on Linux. Even if it shouldn't affect anything, IMO it's worth to test also on other platforms. I generated three PGP keys (`TEST-1`, `TEST-2`, `TEST-3`) to simulate key invalidation. 

  - Download [test-pubkeys.zip](https://github.com/meriadec/ledger-live-desktop/files/2672670/test-pubkeys.zip) and import the three keys:

```bash
$ unzip test-pubkeys.zip -d /tmp/ll-pubkeys
$ gpg --import /tmp/ll-pubkeys/*.asc
```

  - Go to https://github.com/meriadec/ledger-live-desktop/releases
    - here you see:
      - `v2.0.0`, signed with `TEST-1`
      - `v3.0.0`, signed with `TEST-2`
      - `v4.0.0`, signed with `TEST-3`
    - each release contain a `.sha512sum` file, and its signature in `.sha512sum.sig`
- Download *every file* of *every release* (so for each release: the AppImage, the shasum, and the shasum sig)
- Check shasums:

```bash
$ sha512sum -c *.sha512sum                                                                       
ledger-live-desktop-2.0.0-linux-x86_64.AppImage: OK
ledger-live-desktop-3.0.0-linux-x86_64.AppImage: OK
ledger-live-desktop-4.0.0-linux-x86_64.AppImage: OK
```
- Check signatures:

```bash
$ gpg --verify ledger-live-desktop-2.0.0.sha512sum.sig
gpg: assuming signed data in 'ledger-live-desktop-2.0.0.sha512sum'
gpg: Signature made Wed 12 Dec 2018 05:46:46 PM CET
gpg:                using RSA key E96708A0247A58A87E3EBAE6626D563B22D51F51
gpg:                issuer "ll-test-1@ledger.com"
gpg: Good signature from "LL-TEST-1 (Ledger Live Test Key 1) <ll-test-1@ledger.com>" [ultimate]
```
(do same for other `.sig` files, and confirm that `v3.0.0` is signed with `TEST-2`, etc.)

When all is alright, we need to ensure the app is doing the same. For this, two env var to setup: `UPDATE_CHECK_FEED` and `UPDATE_CHECK_IGNORE`.

```bash
$ UPDATE_CHECK_FEED=https://ledger-live.sigsev.io \
  UPDATE_IGNORE_CHECK=0 \
  ./ledger-live-desktop-2.0.0-linux-x86_64.AppImage
```

expected result:

![2018-12-12_517x102](https://user-images.githubusercontent.com/315259/49886045-f5303100-fe38-11e8-880f-caefd4911d50.png)
then:
![2018-12-12_527x111](https://user-images.githubusercontent.com/315259/49886043-f5303100-fe38-11e8-9bd0-823e69ebfdd8.png)

Let's try with incorrect signature:

```bash
$ UPDATE_CHECK_FEED=https://ledger-live.sigsev.io/WRONG-SIG \
  UPDATE_IGNORE_CHECK=0 \
  ./ledger-live-desktop-2.0.0-linux-x86_64.AppImage
```

expected result:

![2018-12-12_597x126](https://user-images.githubusercontent.com/315259/49887935-fdd73600-fe3d-11e8-8025-89fc9be80aee.png)

Additionally, there is a bunch of unit tests, that are ran in the `yarn test` command.

Let's HODL for now until we check all the boxes, keep updated!

:doge:
